pkg/tests/zip_test.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2019 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import filecmp
+import os
+import unittest
+from bazel_tools.tools.python.runfiles import runfiles
+from build_zip import parse_date, ZIP_EPOCH
+from zipfile import ZipFile
+
+HELLO_CRC = 2069210904
+LOREM_CRC = 2178844372
+
+class ZipTest(unittest.TestCase):
+    def get_test_zip(self, zipfile):
+        """Get the file path to a generated zip in the runfiles."""
+
+        return self.data_files.Rlocation(
+            "rules_pkg/tests/" + zipfile
+        )
+
+    def setUp(self):
+        self.data_files = runfiles.Create()
+
+class ZipContentsCase(ZipTest):
+    """Use zipfile to check the contents of some generated zip files."""
+
+    def assertZipFileContent(self, zipfile, content):
+        """Assert that zipfile contains exactly the entries described by
+           content.
+
+        Args:
+            zipfile: the test-package-relative path to a zip file to test.
+            content: an array of dictionaries containing a filename and crc
+                     key, and optionally a timestamp key.
+        """
+        with ZipFile(self.get_test_zip(zipfile)) as f:
+            infos = f.infolist()
+            self.assertEqual(len(infos), len(content))
+
+            for info, expected in zip(infos, content):
```

Comment:
This is going to be sort of brittle, because the content must match the expetected in order. And sometimes we want to just know that the times are correct and others that the names are right. It is fine for now, but what may work more flexibly is something like

                 expect_crc = expected.get('crc')
                 if expect_crc:
                      self.assertEqual(info.CRC, expect_crc)

So any of the fields is optional for the test.   But we can think about that later.

Reply:
Ok